### PR TITLE
Flush before apply if needed

### PR DESCRIPTION
**Pull request by hicqu**
## What have you changed? (mandatory)
Flush the mem table if need before apply snapshot.

## What are the type of the changes? (mandatory)
Improvement.

## How has this PR been tested? (mandatory)
make dev.

## Does this PR affect documentation (docs/docs-cn) update? (mandatory)
No.

## Does this PR affect tidb-ansible update? (mandatory)
No.

### CODE REVIEW

**Comment by siddontang** [comment]
why not flushing before the place where the injection really happens?

I don't think we need to flush every time in DeleteRange

Btw, I worry that this may cause more L0 SST, so maybe it is better to increase the sub-compaction to 2 or more by default.

File: src/raftstore/store/worker/region.rs [hunk 1]
```diff
@@ -279,12 +279,18 @@ impl SnapContext {
...
         check_abort(&abort)?;
         self.cleanup_overlap_ranges(&start_key, &end_key);
-        box_try!(util::delete_all_in_range(
+        let need_flush_mem_table = box_try!(util::delete_all_in_range(
             &self.engines.kv,
             &start_key,
             &end_key,
             self.use_delete_range
         ));
+        for (cf, need_flush) in self.engines.kv.cf_names().iter().zip(need_flush_mem_table) {
```
Comment (zhangjinpeng1987) [inline src/raftstore/store/worker/region.rs]: How about `need_flush_mem_table` is vec![]?

File: src/raftstore/store/util.rs [hunk 2]
```diff
@@ -127,32 +127,37 @@ pub fn conf_change_type_str(conf_type: eraftpb::ConfChangeType) -> &'static str
...
+) -> Result<Vec<bool>> {
     if start_key >= end_key {
-        return Ok(());
+        return Ok(vec![]);
     }
 
-    for cf in db.cf_names() {
-        delete_all_in_range_cf(db, cf, start_key, end_key, use_delete_range)?;
+    let cf_names = db.cf_names();
+    let mut need_flush_mem_table = Vec::with_capacity(cf_names.len());
```
Comment (zhangjinpeng1987) [inline src/raftstore/store/util.rs]: s/need_flush_mem_table/exist_data_before_cleanup
`delete_all_in_range` has nothing to do with flush memtable.

File: src/raftstore/store/util.rs [hunk 3]
```diff
@@ -127,32 +127,37 @@ pub fn conf_change_type_str(conf_type: eraftpb::ConfChangeType) -> &'static str
 
 const MAX_WRITE_BATCH_SIZE: usize = 4 * 1024 * 1024;
 
+/// Return a vector which indicates the cf should be flushed or not for all cfs in the DB.
```
Comment (zhangjinpeng1987) [inline src/raftstore/store/util.rs]: This function has nothing to do with flush memtable.

File: src/raftstore/store/worker/region.rs (hunk 1 above)
Comment (hicqu) [reply src/raftstore/store/worker/region.rs]: So the for loop will be empty.

File: etc/config-template.toml [hunk 4]
```diff
@@ -228,7 +228,7 @@
 # This value represents the maximum number of threads that will concurrently perform a
 # compaction job by breaking it into multiple, smaller ones that are run simultaneously.
 # Default: 1 (i.e. no subcompactions)
-# max-sub-compactions = 1
+max-sub-compactions = 2
```
Comment (siddontang) [inline etc/config-template.toml]: we also need to change the default value in Config structure.

Comment (hicqu) [reply etc/config-template.toml]: oh, it's better than change it here. So I will keep the line commented.

File: src/raftstore/store/util.rs [hunk 5]
```diff
@@ -164,36 +164,43 @@ pub fn delete_all_in_range(
...
     end_key: &[u8],
     use_delete_range: bool,
-) -> Result<()> {
+) -> Result<bool> {
     let handle = rocksdb_util::get_cf_handle(db, cf)?;
     let mut wb = WriteBatch::new();
-    // Since CF_RAFT and CF_LOCK is usually small, so using
-    // traditional way to cleanup.
+
+    let mut exist_data_before_cleanup = false;
```
Comment (siddontang) [inline src/raftstore/store/util.rs]: em, seem that mostly, this will be set to true?

File: etc/config-template.toml [hunk 6]
```diff
@@ -228,7 +228,7 @@
 # This value represents the maximum number of threads that will concurrently perform a
 # compaction job by breaking it into multiple, smaller ones that are run simultaneously.
 # Default: 1 (i.e. no subcompactions)
```
Comment (MyonKeminta) [inline etc/config-template.toml]: Please also update the default value in the comment

**Comment by huachaohuang** [comment]
Let me take over this problem by utilizing [this option](https://github.com/facebook/rocksdb/blob/master/include/rocksdb/options.h#L1268) and [this commit](https://github.com/facebook/rocksdb/pull/4297).

**Comment by Hoverbear** [comment]
@huachaohuang Will you take over this PR, or open a new one? If you will make a new one maybe we can close this?

**Comment by huachaohuang** [comment]
Close this in favor of https://github.com/pingcap/rust-rocksdb/pull/218.

**Pull request closed without merging**